`ActiveModel::Serializers::JSON#as_json` does deep `as_json`

tamird opened the pull request:
This corrects a bug where `#as_json` on `ActiveModel`s with date fields
does not return a hash of primitives, as would be expected. That is,
prior to this change, `#as_json` would return a hash of Ruby objects
that do not all have direct corollaries in JSON.

rebirth of https://github.com/rails/rails/pull/17132

tamird commented:
@strzalek @chancancode ping

strzalek commented:
Looks good. I'm :+1: on this.

tamird commented:
@rafaelfranca mind triaging this?

rafaelfranca commented:
I'll leave to @chancancode, he knows more about this

chancancode commented:
Sorry, I'll be quite busy for the next week or so, so I'll be a bit slow at responding :sa: 

I still don't understand the need for this. We consider `as_json` encoder hooks and we don't guarantee anything about their return values. Is there a case where `to_json` would return different result before/after this change?

tamird commented:
> We consider as_json encoder hooks and **we don't guarantee anything about their return values**.

Can you show me where this is codified? My understand of the semantics of `as_json` is that it is guaranteed to return primitive objects that can be expressed as JSON - e.g. symbols are converted to strings. As mentioned in the original description, this change solves the case of `ActiveModel#as_json` on objects with `Date` fields returning a hash with `Date` values, rather than strings. JSON does not have a `Date` type.

chancancode commented:
> My understand of the semantics of as_json is that it is guaranteed to return primitive objects that can be expressed as JSON - e.g. symbols are converted to strings.

Yeah, this is what I am saying – this is not the case, by design. Or at least the encoder does not *require* `as_json` to do that. The design is that you can *choose to* override `as_json` to hint the encoder about how you want your object represented.

We feel that it would be too much of a burden if we *require* developers to return only primitives, as they would have to worry about walking the tree recursively, etc (it's not trivial, there are a lot of edge cases).

So, you can basically return anything in `as_json` (and that happens all the time, and we have many test cases for those). If you return a primitive, the encoder won't touch it. If you return something that's not a primitive (e.g. a symbol), the encoder will encode it however it deems to make sense.

For example, [this implementation](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/json.rb#L193-L197) returns symbols in the hash. *Could* it have returned it in string keys? Sure.  But it is not "wrong" to return symbol keys either. The only thing [we care](https://github.com/rails/rails/blob/master/activesupport/test/json/encoding_test.rb#L132-L137) is the `to_json` result.

So, perhaps `as_json` is a misleading name, and it should have been `for_json` or something, but it's too late to change :/

> Can you show me where this is codified?

I agree that the documentation on this is very spotting, sometimes inaccurate even :neutral_face: I had to trace through a lot of git history (and had a lot of help from @jeremy) to learn all that. We should probably fix this!

So, of course, I know that people "rely" on `as_json` all the time, and we *could* always "go an extra mile" and deeply jsonify things. It would still be "correct", because it doesn't change the result. But we have to draw a line somewhere, because that's just not how we designed it to work. And in the long run, this would hurt performance (or at least hinder our ability to improve it) because we would have to do two recursive tree walks – one in `as_json` to "be nice", and once the encoder because we must.

Does that help?

tamird commented:
Yeah, I understand what you're saying - this patch is not intended to globally fix `as_json`'s behaviour, it merely changes ActiveModel's `as_json` to be better-behaved with respect to returning primitives. The additional cost in performance is one additional method call per field - is that enough to reject this improvement?

tamird committed:
`ActiveModel::Serializers::JSON#as_json` does deep `as_json`

This corrects a bug where `#as_json` on `ActiveModel`s with date fields
does not return a hash of primitives, as would be expected. That is,
prior to this change, `#as_json` would return a hash of Ruby objects
that do not all have direct corollaries in JSON.

tamird commented:
Reborn as #22143.
